automation/publish_signals.go

```diff
@@ -0,0 +1,268 @@
+	_ = enc.Encode(v)
+}
+
+func (opts LogOptions) Printf(format string, a ...any) {
+	if opts.Out == nil {
+		return
+	}
+	fmt.Fprintf(opts.Out, format, a...)
+}
+
+// PublishOptions describe options that can be supplied when running a
+// PublishRuleSet.
+type PublishOptions struct {
+	LogOptions
+
+	// DryRun, if true, does not run the automation, but rather logs what it's
+	// planning to do.
+	DryRun bool
+
+	// Publisher is a name describing the publisher application. The default is
+	// the declared path of the main module.
+	Publisher string
+}
+
+func (opts PublishOptions) withDefaults() PublishOptions {
+	if opts.Publisher == "" {
+		info, ok := debug.ReadBuildInfo()
+		if ok {
+			opts.Publisher = info.Main.Path
+		}
+	}
+	return opts
+}
+
+// PublishSignals allows you to automate signal publishing from one or more
+// source integrations.
+type PublishSignals struct {
+	// Integrations must list the IDs of the integrations to publish signals
+	// from. If this list is empty, the rule set is a no-op.
+	Integrations []string
+
+	// SignalsFilter can optionally be specified to limit which signals to
+	// publish.
+	SignalsFilter query.FilterType
+
+	// TransformVersion should be changed if you want existing items to be
+	// republished despite the source signal being unchanged.
+	TransformVersion string
+
+	// Transforms is a list of transforms to apply when publishing the signals.
+	// The transforms are applied in order.
+	Transforms []func(item *views.ItemSave)
+}
+
+// Do performs the automation against c with the passed in opts.
+func (p PublishSignals) Do(ctx context.Context, c *clarify.Client, opts PublishOptions) error {
+	var total int
+	defer func() {
+		var suffix string
+		if opts.DryRun {
+			suffix = " (dry-run)"
+		}
+		opts.Printf("-- Published %d signals from %d integrations%s.\n", total, len(p.Integrations), suffix)
+	}()
+
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	opts = opts.withDefaults()
+
+	// We iterate signals without requesting the total count. This is an
+	// optimization bet that total % limit == 0 is uncommon.
+	q := query.Query{
+		Filter: p.SignalsFilter.Filter(),
+		Sort:   []string{"id"},
+		Limit:  selectSignalsPageSize,
+	}
+
+	items := make(map[string]views.ItemSave)
+	var flush func(string) error
+	if opts.DryRun {
+		flush = func(integrationID string) error {
+			opts.Printf("Publish %d items (dry-run)...\n", len(items))
+			if opts.Verbose {
+				opts.EncodeJSON(items)
+			}
+			total += len(items)
+			maps.Clear(items)
+			return nil
+		}
+	} else {
+		flush = func(integrationID string) error {
+			opts.Printf("Publish %d items...\n", len(items))
+			if opts.Verbose {
+				opts.Printf("itemsBySignal:\n")
+				opts.EncodeJSON(items)
+			}
+			result, err := c.PublishSignals(integrationID, items).Do(ctx)
+			if err != nil {
+				return err
+			}
+			if opts.Verbose {
+				opts.EncodeJSON(result)
+			}
+			total += len(items)
+			maps.Clear(items)
+			return nil
+		}
+	}
```

Comment:
We have code duplication for `flush` function, maybe better to go with next option?
```suggestion
flush = func(integrationID string) error {
	opts.Printf("Publish %d items...\n", len(items))
	if opts.Verbose {
		opts.Printf("itemsBySignal:\n")
		opts.EncodeJSON(items)
	}

	if !opts.DryRun {
		result, err := c.PublishSignals(integrationID, items).Do(ctx)
		if err != nil {
			return err
		}
		if opts.Verbose {
			opts.EncodeJSON(result)
		}
	}
	
	total += len(items)
	maps.Clear(items)
	return nil
}
```

Reply:
Sure, can go with that for this case.

The reason I have duplication is to reduce branches (if statement) within the for loop, but it's a bit overkill here as it's not a "tight" loop.

Will change it.

Reply:
fixed